File: ex975.cpp
// DP, ordered_map: key=>the value at position i; value=>the next index

class Solution {
public:
    int oddEvenJumps(vector<int>& A) {
        int n = A.size();
        vector<bool> dp_odd(n);
        vector<bool> dp_even(n);
        dp_odd[n - 1] = 1;    dp_even[n - 1] = 1;
        
        // Init treemap
        map<int, int> treemap;
        treemap[A[n - 1]] = n - 1;
        
        // Build dp arrays
        for (int i = n - 2; i >= 0; --i) {
            int key = A[i];
            if (treemap.find(key) != treemap.end()) {
                dp_odd[i] = dp_even[treemap[key]];
                dp_even[i] = dp_odd[treemap[key]];
            } else {
                auto iter1 = treemap.lower_bound(A[i]);   
                if (iter1 != treemap.end())
                {                
                    if (dp_even[iter1->second]) dp_odd[i] = true;
                }
            
                auto iter2 = treemap.upper_bound(A[i]);
                if (iter2 != treemap.begin())
                {
                    iter2 = prev(iter2,1);                
                    if (dp_odd[iter2->second]) dp_even[i] = true;
                }
            }
            treemap[key] = i;
        }
        
        // Get result
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            cnt += dp_odd[i];
        }
        return cnt;
    }
};

File: ex975.java
class Solution {
    public int oddEvenJumps(int[] A) {
        int n = A.length;
        if (n <= 1)    return n;
        
        // Build dp array
        boolean[] dp_odd = new boolean[n];
        boolean[] dp_even = new boolean[n];
        dp_odd[n - 1] = true;    dp_even[n - 1] = true;
        
        // Build treemap
        TreeMap<Integer, Integer> treemap = new TreeMap<>();
        treemap.put(A[n - 1], n - 1);
        
        for (int i = n - 2; i >= 0; i--) {
            Integer odd_next = treemap.ceilingKey(A[i]);
            Integer even_next = treemap.floorKey(A[i]);

            dp_odd[i] = odd_next == null ? false : dp_even[treemap.get(odd_next)];
            dp_even[i] = even_next == null ? false : dp_odd[treemap.get(even_next)];
            
            treemap.put(A[i], i);
        }
        
        // Get results
        int res = 0;
        for (int i = 0; i < n; i++)
            if (dp_odd[i])    res++;
        
        return res;
    }
}